File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/BT/GraphColouring.swift
import Foundation

func printSolution(_ colour : inout [Int], _ V : Int) {
	print("Assigned colours are::", terminator: "")
	var i : Int = 0
	while (i < V) {
		print(String(colour[i]), terminator: " ")
		i += 1
	}
	print()
}

// Check if the whole graph is coloured properly.
func isSafe2(_ graph : inout [[Bool]], _ colour : inout [Int], _ V : Int) -> Bool {
		var i : Int = 0
		while (i < V) {
			var j : Int = i + 1
			while (j < V) {
				if (graph[i][j] && colour[j] == colour[i]) {
					return false
				}
				j += 1
			}
			i += 1
		}
	return true
}

func graphColouring2(_ graph : inout [[Bool]], _ V : Int, _ m : Int, _ colour : inout [Int], _ i : Int) -> Bool {
	if (i == V) {
		if (isSafe2( &graph, &colour,V)) {
			printSolution( &colour,V)
			return true
		}
		return false
	}
	var j : Int = 1
	// Assign each colour from 1 to m
	while (j <= m) {
		colour[i] = j
		if (graphColouring2( &graph,V,m, &colour,i + 1)) {
			return true
		}
		j += 1
	}
	return false
}

func graphColouring2(_ graph : inout [[Bool]], _ V : Int, _ m : Int) -> Bool {
	var colour : [Int] = Array(repeating: 0, count: V)
	if (graphColouring2( &graph,V,m, &colour,0)) {
		return true
	}
	return false
}

// Is it safe to colour vth vertice with c colour.
func isSafe(_ graph : inout [[Bool]], _ V : Int, _ colour : inout [Int], _ v : Int, _ c : Int) -> Bool {
	var i : Int = 0
	while (i < V) {
		if (graph[v][i] == true && c == colour[i]) {
			return false
		}
		i += 1
	}
	return true
}

func graphColouringUtil(_ graph : inout [[Bool]], _ V : Int, _ m : Int, _ colour : inout [Int], _ i : Int) -> Bool {
	if (i == V) {
		printSolution( &colour,V)
		return true
	}
	var j : Int = 1
	while (j <= m) {
		if (isSafe( &graph,V, &colour,i,j)) {
			colour[i] = j
			if (graphColouringUtil( &graph,V,m, &colour,i + 1)) {
				return true
			}
		}
		j += 1
	}
	return false
}

func graphColouring(_ graph : inout [[Bool]], _ V : Int, _ m : Int) -> Bool {
	var colour : [Int] = Array(repeating: 0, count: V)
	if (graphColouringUtil( &graph,V,m, &colour,0)) {
		return true
	}
	return false
}

// Testing code.
var graph : [[Bool]] =
[[false, true, false, false, true],
[true, false, true, false, true],
[false, true, false, true, true],
[false, false, true, false, true],
[true, true, true, true, false]]

let V : Int = 5 // Number of vertices
let m : Int = 4 // Number of colours

if (!graphColouring2( &graph,V,m)) {
	print("Solution does not exist")
}
if (!graphColouring( &graph,V,m)) {
	print("Solution does not exist")
}

/* 
Assigned colours are::1 2 1 2 3 
Assigned colours are::1 2 1 2 3  
*/

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/BT/GraphColouring.ts
// Check if the whole graph is coloured properly.
function isSafe2(graph: boolean[][], colour: number[], V: number): boolean {
    for (let i = 0; i < V; i++) {
        for (let j = i + 1; j < V; j++) {
            if (graph[i][j] && colour[j] === colour[i]) {
                return false;
            }
        }
    }
    return true;
}

function graphColouring2Util(graph: boolean[][], V: number, m: number, colour: number[], i: number): boolean {
    if (i === V) {
        if (isSafe2(graph, colour, V)) {
            console.log("Assigned colours are: " + colour.join(", "));
            return true;
        }
        return false;
    }

    for (let j = 1; j <= m; j++) {
        colour[i] = j;
        if (graphColouring2Util(graph, V, m, colour, i + 1)) {
            return true;
        }
    }
    return false;
}

function graphColouring2(graph: boolean[][], V: number, m: number): boolean {
    const colour: number[] = new Array(V).fill(0);
    if (graphColouring2Util(graph, V, m, colour, 0)) {
        return true;
    }
    return false;
}

// Is it safe to colour vth vertice with c colour.
function isSafe(graph: boolean[][], V: number, colour: number[], v: number, c: number): boolean {
    for (let i = 0; i < V; i++) {
        if (graph[v][i] && c === colour[i]) {
            return false;
        }
    }
    return true;
}

function graphColouringUtil(graph: boolean[][], V: number, m: number, colour: number[], i: number): boolean {
    if (i === V) {
        console.log("Assigned colours are: " + colour.join(", "));
        return true;
    }

    for (let j = 1; j <= m; j++) {
        if (isSafe(graph, V, colour, i, j)) {
            colour[i] = j;
            if (graphColouringUtil(graph, V, m, colour, i + 1)) {
                return true;
            }
        }
    }
    return false;
}

function graphColouring(graph: boolean[][], V: number, m: number): boolean {
    const colour: number[] = new Array(V).fill(0);
    if (graphColouringUtil(graph, V, m, colour, 0)) {
        return true;
    }
    return false;
}

const graph: boolean[][] = [
    [false, true, false, false, true],
    [true, false, true, false, true],
    [false, true, false, true, true],
    [false, false, true, false, true],
    [true, true, true, true, false]
];

const V: number = 5; // Number of vertices
const m: number = 4; // Number of colours

if (!graphColouring2(graph, V, m)) {
    console.log("Solution does not exist.");
}

if (!graphColouring(graph, V, m)) {
    console.log("Solution does not exist.");
}

/*
Assigned colours are: 1, 2, 1, 2, 3
Assigned colours are: 1, 2, 1, 2, 3
*/
